## Data Collection & Preparation

* Pandas for managing the data
* NumPy for math operations
* Sklearn for Machine Learning
* Seaborn for advanced visualization
* Matplotlib for additional plottingtinging

In [1]:
# Import libraries 

import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_squared_error, r2_score
import warnings
warnings.filterwarnings("ignore", category=UserWarning) 
%matplotlib inline

In [7]:
#read data files

cs = pd.read_csv("charging_sessions.csv")
wba = pd.read_csv("weather_burbank_airport.csv")

In [8]:
#print the first 5 rows of the charging_sessions dataset
cs.head()

Unnamed: 0                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...

In [9]:
#print out the first 5 rows of the weather burbank airport dataset
wba.head()

city            timestamp  temperature  cloud_cover  \
0  Burbank  2018-01-01 08:53:00          9.0         33.0   
1  Burbank  2018-01-01 09:53:00          9.0         33.0   
2  Burbank  2018-01-01 10:53:00          9.0         21.0   
3  Burbank  2018-01-01 11:53:00          9.0         29.0   
4  Burbank  2018-01-01 12:53:00          8.0         33.0   

  cloud_cover_description  pressure  windspeed  precipitation  \
0                    Fair    991.75        9.0            0.0   
1                    Fair    992.08        0.0            0.0   
2                    Haze    992.08        0.0            0.0   
3           Partly Cloudy    992.08        0.0            0.0   
4                    Fair    992.08        0.0            0.0   

   felt_temperature  
0               8.0  
1               9.0  
2               9.0  
3               9.0  
4               8.0

In [12]:
#print out the columns of charging sessions dataset
cs.columns

Index(['Unnamed: 0', 'id', 'connectionTime', 'disconnectTime',
       'doneChargingTime', 'kWhDelivered', 'sessionID', 'siteID', 'spaceID',
       'stationID', 'timezone', 'userID', 'userInputs'],
      dtype='object')

In [13]:
#print out the columns of weather burbank airport dataset
wba.columns

Index(['city', 'timestamp', 'temperature', 'cloud_cover',
       'cloud_cover_description', 'pressure', 'windspeed', 'precipitation',
       'felt_temperature'],
      dtype='object')

In [14]:
#print out the tail of cs dataset to see how many entries there are
cs.tail()

Unnamed: 0                        id             connectionTime  \
66445       10083  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446       10084  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447       10085  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448       10086  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   
66449       10087  5d574ad2f9af8b4c10c03656  2019-07-31 19:21:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   
66449  2019-07-31 22:00:04+00:00  2019-07-31 20:51:34+00:00         0.897   

                                    sessionID  siteID  spaceID    stationID  \
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697       1  AG-1F09  1-1-178-817   

                  timezone  userID  \
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   
66449  America/Los_Angeles     NaN   

                                              userInputs  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  
66449                                                NaN

In [15]:
#print out the tail of cs dataset to see how many entries there are
wba.tail()

city            timestamp  temperature  cloud_cover  \
29239  Burbank  2021-01-01 03:53:00         13.0         33.0   
29240  Burbank  2021-01-01 04:53:00         12.0         33.0   
29241  Burbank  2021-01-01 05:53:00         12.0         33.0   
29242  Burbank  2021-01-01 06:53:00         11.0         33.0   
29243  Burbank  2021-01-01 07:53:00         10.0         33.0   

      cloud_cover_description  pressure  windspeed  precipitation  \
29239                    Fair    986.81        0.0            0.0   
29240                    Fair    986.81       11.0            0.0   
29241                    Fair    987.47        9.0            0.0   
29242                    Fair    987.14       13.0            0.0   
29243                    Fair    987.80        6.0            0.0   

       felt_temperature  
29239              13.0  
29240              12.0  
29241              12.0  
29242              11.0  
29243              10.0

### We will focus on the charging_sessions dataset for a while

In [16]:
#print information about cs dataset
cs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


We can see that this dataset has an unnamed column that represents the entry number. For better readability, we will rename this column

In [25]:
#rename the unnamed column
cs.rename( columns={'Unnamed: 0':'number'}, inplace=True )

In [19]:
#check if the column was renamed
cs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   number            66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


From looking at the count of every column, we can see that the "doneChargingTime" is missing values. The missing values in the "userID" and "userInputs" columns are because the user was not registered, so we can ignore them for now.

**Detecting missing numerical data**

In [24]:
#check if we can spot null-values
cs.isnull()

number     id  connectionTime  disconnectTime  doneChargingTime  \
0       False  False           False           False             False   
1       False  False           False           False             False   
2       False  False           False           False             False   
3       False  False           False           False             False   
4       False  False           False           False             False   
...       ...    ...             ...             ...               ...   
66445   False  False           False           False             False   
66446   False  False           False           False             False   
66447   False  False           False           False             False   
66448   False  False           False           False             False   
66449   False  False           False           False             False   

       kWhDelivered  sessionID  siteID  spaceID  stationID  timezone  userID  \
0             False      False   False    False      False     False   False   
1             False      False   False    False      False     False   False   
2             False      False   False    False      False     False   False   
3             False      False   False    False      False     False   False   
4             False      False   False    False      False     False   False   
...             ...        ...     ...      ...        ...       ...     ...   
66445         False      False   False    False      False     False   False   
66446         False      False   False    False      False     False   False   
66447         False      False   False    False      False     False   False   
66448         False      False   False    False      False     False   False   
66449         False      False   False    False      False     False    True   

       userInputs  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
66445       False  
66446       False  
66447       False  
66448       False  
66449        True  

[66450 rows x 13 columns]

In [30]:
#drop all rows containing null values
cs.dropna(axis=0, inplace=False)

number                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
...       ...                       ...                        ...   
66444   10082  5d574ad2f9af8b4c10c03651  2019-07-31 14:50:17+00:00   
66445   10083  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446   10084  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447   10085  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448   10086  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
0      2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1      2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2      2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3      2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4      2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
...                          ...                        ...           ...   
66444  2019-08-01 01:01:18+00:00  2019-07-31 20:29:24+00:00        27.174   
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
66444  1_1_191_784_2019-07-31 14:50:17.037367       1  AG-4F40  1-1-191-784   
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
66444  America/Los_Angeles   448.0   
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   

                                              userInputs  
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...  
...                                                  ...  
66444  [{'WhPerMile': 200, 'kWhRequested': 28.0, 'mil...  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  

[46008 rows x 13 columns]

In [31]:
#check how many rows are getting dropped
len(cs)-len(cs.dropna())

20442

This number is higher than expected, but because this factors in the null values from userID and userInputs(unregistered users or users with no Input), we want so specifically look for missing values in the "doneChargingTime"-column

In [32]:
len(cs)-len(cs.dropna(subset = ['doneChargingTime']))

4088

This number tells us that there are 4088 sessions with no timestamp for when the EV got done charging, meaning that it didn't fully charge. Let's look at this data.

In [44]:
#display records with 'doneCharging = Null
cs_notdone = cs[cs['doneChargingTime'].isnull()]
cs_notdone

number                        id             connectionTime  \
3756     3756  5ea3848df9af8b428bf97225  2020-04-08 16:36:53+00:00   
3757     3757  5ea4d60df9af8b46573f1ee7  2020-04-09 13:23:52+00:00   
3758     3758  5ea4d60df9af8b46573f1ee8  2020-04-09 13:40:03+00:00   
3759     3759  5ea4d60df9af8b46573f1ee9  2020-04-09 14:01:53+00:00   
3760     3760  5ea4d60df9af8b46573f1eea  2020-04-09 15:01:06+00:00   
...       ...                       ...                        ...   
60411    4049  5cb3d2a8f9af8b4551261e75  2019-03-30 00:32:10+00:00   
61097    4735  5cc8e891f9af8b45d975b819  2019-04-14 16:19:12+00:00   
61103    4741  5cca3a54f9af8b49aaa4cbac  2019-04-15 13:00:59+00:00   
61117    4755  5cca3a54f9af8b49aaa4cbba  2019-04-15 13:52:45+00:00   
65530    9168  5d3f8f96f9af8b7f97666a55  2019-07-13 20:13:18+00:00   

                  disconnectTime doneChargingTime  kWhDelivered  \
3756   2020-04-09 03:43:29+00:00              NaN        25.183   
3757   2020-04-09 17:59:02+00:00              NaN        14.136   
3758   2020-04-09 21:43:22+00:00              NaN        14.203   
3759   2020-04-09 16:33:40+00:00              NaN        13.041   
3760   2020-04-09 23:38:08+00:00              NaN         5.797   
...                          ...              ...           ...   
60411  2019-03-30 02:23:14+00:00              NaN        11.148   
61097  2019-04-14 20:38:39+00:00              NaN        11.678   
61103  2019-04-15 13:55:25+00:00              NaN         3.343   
61117  2019-04-15 14:47:01+00:00              NaN         4.558   
65530  2019-07-13 20:37:40+00:00              NaN         2.218   

                                    sessionID  siteID  spaceID    stationID  \
3756   1_1_194_822_2020-04-08 16:36:53.316264       1  AG-1F12  1-1-194-822   
3757   1_1_193_827_2020-04-09 13:23:51.689520       1  AG-1F02  1-1-193-827   
3758   1_1_178_823_2020-04-09 13:39:55.115497       1  AG-1F08  1-1-178-823   
3759   1_1_178_828_2020-04-09 14:01:53.467693       1  AG-1F10  1-1-178-828   
3760   1_1_193_820_2020-04-09 15:01:06.343454       1  AG-1F04  1-1-193-820   
...                                       ...     ...      ...          ...   
60411  1_1_193_829_2019-03-30 00:32:10.128651       1  AG-1F03  1-1-193-829   
61097  1_1_193_816_2019-04-14 16:19:11.864262       1  AG-1F05  1-1-193-816   
61103  1_1_193_819_2019-04-15 13:00:58.722284       1  AG-1F06  1-1-193-819   
61117  1_1_179_800_2019-04-15 13:52:44.693153       1  AG-3F32  1-1-179-800   
65530  1_1_193_825_2019-07-13 20:13:18.490314       1  AG-1F01  1-1-193-825   

                  timezone  userID  \
3756   America/Los_Angeles   507.0   
3757   America/Los_Angeles   419.0   
3758   America/Los_Angeles  5065.0   
3759   America/Los_Angeles   651.0   
3760   America/Los_Angeles   466.0   
...                    ...     ...   
60411  America/Los_Angeles   194.0   
61097  America/Los_Angeles   364.0   
61103  America/Los_Angeles   651.0   
61117  America/Los_Angeles   194.0   
65530  America/Los_Angeles    76.0   

                                              userInputs  
3756   [{'WhPerMile': 250, 'kWhRequested': 75.0, 'mil...  
3757   [{'WhPerMile': 400, 'kWhRequested': 40.0, 'mil...  
3758   [{'WhPerMile': 400, 'kWhRequested': 24.0, 'mil...  
3759   [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...  
3760   [{'WhPerMile': 300, 'kWhRequested': 6.0, 'mile...  
...                                                  ...  
60411  [{'WhPerMile': 400, 'kWhRequested': 24.0, 'mil...  
61097  [{'WhPerMile': 400, 'kWhRequested': 56.0, 'mil...  
61103  [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...  
61117  [{'WhPerMile': 400, 'kWhRequested': 12.0, 'mil...  
65530  [{'WhPerMile': 301, 'kWhRequested': 9.03, 'mil...  

[4088 rows x 13 columns]

### Dealing with missing values

When dealing with missing values, we can either **eliminate** them from the dataset or **impute** the null values with estimates. Because the missing data points are timestamps, it's hard them because they simply do not exist. The EV never got done charging, so estimating a value doesn't make any sense. And because our job is to optimize utilization, looking at data points where the EV never fully charged(and thus didn't stay plugged in without charging and taking up charging space) is not adequate. That is why we will drop these data points.

In [45]:
#we take every row where "doneChargingTime" is not null into our new dataset cs_clean
cs_clean = cs[cs['doneChargingTime'].notna()]

In [46]:
cs_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 62362 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   number            62362 non-null  int64  
 1   id                62362 non-null  object 
 2   connectionTime    62362 non-null  object 
 3   disconnectTime    62362 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      62362 non-null  float64
 6   sessionID         62362 non-null  object 
 7   siteID            62362 non-null  int64  
 8   spaceID           62362 non-null  object 
 9   stationID         62362 non-null  object 
 10  timezone          62362 non-null  object 
 11  userID            46008 non-null  float64
 12  userInputs        46008 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.7+ MB


The data is now clean and we can work with it!